File: 백준/Gold/15971. 두 로봇/두 로봇.cc
/**
 * Author: nowalex322, Kim HyeonJae
 */
#include <bits/stdc++.h>
using namespace std;

// #define int long long
#define MOD 1000000007
#define INF LLONG_MAX
#define ALL(v) v.begin(), v.end()

#ifdef LOCAL
#include "algo/debug.h"
#else
#define debug(...) 42
#endif

struct Node {
    int to;
    int W;
    Node(int to, int W) : to(to), W(W) {}
};

vector<vector<Node>> board;
vector<bool> visited;
int N, robot1, robot2, totalD, maxW;

bool dfs(int start, int end) {
    if (start == end) return true;

    visited[start] = true;

    for (const Node& n : board[start]) {
        if (!visited[n.to]) {
            int prevMaxW = maxW;
            maxW = max(maxW, n.W);
            totalD += n.W;

            if (dfs(n.to, end)) return true;

            maxW = prevMaxW;
            totalD -= n.W;
        }
    }

    visited[start] = false;
    maxW = 0;
    return false;
}

void solve() {
    cin >> N >> robot1 >> robot2;
    if (N == 1 || robot1 == robot2) {
        cout << 0 << "\n";
        return;
    }

    board.resize(N + 1);
    visited.resize(N + 1, false);

    for (int i = 0; i < N - 1; i++) {
        int first, second, w;
        cin >> first >> second >> w;
        board[first].push_back(Node(second, w));
        board[second].push_back(Node(first, w));
    }

    totalD = 0;
    maxW = 0;

    dfs(robot1, robot2);
    cout << totalD - maxW << "\n";
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int tt = 1;  // 기본적으로 1번의 테스트 케이스를 처리
    // cin >> tt;    // 테스트 케이스 수 입력 (필요 시)

    while (tt--) {
        solve();
    }
    return 0;
}

File: 백준/Gold/15971. 두 로봇/두 로봇.java
/**
 * Author: nowalex322, Kim HyeonJae
 */
import java.io.*;
import java.util.*;

public class Main {
	class Node{
		int to;
		int W;
		public Node(int to, int W){
			this.to = to;
			this.W = W;
		}
	}
	static BufferedReader br;
	static BufferedWriter bw;
	static StringTokenizer st;
	static int N, robot1, robot2, first, second, w, totalD, maxW;
	static ArrayList<Node>[] map;
	static boolean[] visited;
	public static void main(String[] args) throws Exception {
		new Main().solution();
	}

	public void solution() throws Exception {
		br = new BufferedReader(new InputStreamReader(System.in));
//		br = new BufferedReader(new InputStreamReader(new FileInputStream("input.txt")));
		bw = new BufferedWriter(new OutputStreamWriter(System.out));
		st = new StringTokenizer(br.readLine());

		N = Integer.parseInt(st.nextToken());
		robot1 = Integer.parseInt(st.nextToken());
		robot2 = Integer.parseInt(st.nextToken());

		map = new ArrayList[N+1];
		for(int i=1; i<=N; i++) {
			map[i] = new ArrayList<Node>();
		}
		
		for(int i=0; i<N-1; i++) {
			st = new StringTokenizer(br.readLine());
			first = Integer.parseInt(st.nextToken());
			second = Integer.parseInt(st.nextToken());
			w = Integer.parseInt(st.nextToken());
			
			map[first].add(new Node(second, w));
			map[second].add(new Node(first, w));
		}
		
		visited = new boolean[N+1];
		totalD = 0;
		maxW = 0;
		
		dfs(robot1, robot2, 0);
		bw.write(String.valueOf(totalD - maxW));
		bw.flush();
		bw.close();
		br.close();
	}

	private boolean dfs(int start, int end, int sum) {
		if(start == robot2) {
			totalD = sum;
			return true;
		}
		
		visited[start] = true;
		
		for(Node n : map[start]) {
			if(!visited[n.to]) {
				maxW = Math.max(maxW, n.W);
				if(dfs(n.to, end, sum + n.W)) return true;
			}
		}
		return false;
	}
}